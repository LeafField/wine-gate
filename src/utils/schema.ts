import { z } from "zod";

export const searchSchema = z.object({
  search: z
    .string()
    .min(4, { message: "4文字以上入力してください" })
    .max(60, { message: "60文字以下で入力してください" }),
  category: z.string(),
  sort: z.enum(["new", "inexpensive", "popular"]),
});
