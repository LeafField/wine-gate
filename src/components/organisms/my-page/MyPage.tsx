import React, { FC } from "react";
import LineUpHero from "../../molecules/lineup-hero/LineUpHero";
import WineTitle from "../../atoms/wine-title/WineTitle";
import { WineTitleProps } from "../../../types/article_types";
import BackLink from "../../atoms/back-link/BackLink";
import TopPageLink from "../../atoms/toppage-link/TopPageLink";

type Props = {
  heroTitle: string;
  pageTitle: string;
  articles: WineTitleProps[];
};

const MyPage: FC<Props> = ({ heroTitle, pageTitle, articles }) => {
  return (
    <>
      <LineUpHero title={heroTitle} />
      <h3 className="mx-4 border-b-2 border-red text-heading2 sort tablet:mx-0">
        {pageTitle}
      </h3>
      <main className="max-w-[64.9375rem] space-y-4 px-4 main tablet:w-[calc(var(--vw)-401px)] tablet:px-0">
        {articles.map((article) => (
          <WineTitle key={article.id} id={article.id} title={article.title} />
        ))}
      </main>
      <div className="footer-area">
        <BackLink />
        <TopPageLink />
      </div>
    </>
  );
};

export default MyPage;
