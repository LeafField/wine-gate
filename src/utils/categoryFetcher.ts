import { supabase } from "./supabase";
import { cache } from "react";
import { ArticleProps } from "../types/article_types";

export const categoryFetcher = cache(
  async (slug: string[]): Promise<ArticleProps[] | null> => {
    if (slug.length !== 3) return null;
    const category = slug[0];
    const sort = slug[1];
    const page = Number(slug[2]);
    const start = (page - 1) * 10;
    const end = page * 10 - 1;
    if (sort === "new") {
      const { data, error } = await supabase
        .from("wines")
        .select(
          "title,sober_or_sweet,tart,fruity,author_name,tags,id,price,image_src,categories(*)",
        )
        .eq("categories.category", category)
        .order("created_at", { ascending: false })
        .range(start, end);
      if (error) throw new Error("カテゴリの取得に失敗しました");
      return data;
    } else if (sort === "inexpensive") {
      const { data, error } = await supabase
        .from("wines")
        .select(
          "title,sober_or_sweet,tart,fruity,author_name,tags,id,price,image_src,categories(*)",
        )
        .eq("categories.category", category)
        .order("price", { ascending: true })
        .range(start, end);
      if (error) throw new Error("カテゴリの取得に失敗しました");
      return data;
    } else if (sort === "popular") {
      const { data, error } = await supabase
        .rpc("get_wine_category_favorite", {
          category_value: category,
        })
        .limit(10);
      if (error) throw new Error("カテゴリの取得に失敗しました");
      const articles: ArticleProps[] = data.map((article) => ({
        categories: {
          category: article.category_name,
          id: article.category_id,
          chara: article.category_chara,
          sub: article.category_sub,
        },
        author_name: article.author_name,
        fruity: article.fruity,
        id: article.id,
        image_src: article.image_src,
        price: article.price,
        sober_or_sweet: article.sober_or_sweet,
        tags: article.tags,
        tart: article.tart,
        title: article.title,
      }));

      return articles;
    }
    return null;
  },
);

export const categoryCountFetcher = cache(async (category: string) => {
  const { count } = await supabase
    .from("wines")
    .select("id,categories(category)", { count: "exact", head: true })
    .eq("categories.category", category);
  if (count) {
    const totalPage = Math.ceil(count / 10);
    return totalPage;
  } else {
    return 1;
  }
});
