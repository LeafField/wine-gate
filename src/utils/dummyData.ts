import dummyImage from "../images/dummy_wine.png";
import {
  SmallArticleProps,
  CategoryProps,
  WineProps,
  ArticleProps,
} from "../types/article_types";

export const smallArticleDummyData: SmallArticleProps[] = [
  {
    id: "1",
    title: "コノスル カベルネソーヴィニヨン ビシクレタ・レゼルバ",
    chara: "渋い",
    tags: "樽熟成",
    image_src: dummyImage.src,
  },
  {
    id: "2",
    title: "コノスル シャルドネ ビシクレタ・レゼルバ",
    chara: "辛口白",
    tags: "フルーティ",
    image_src: dummyImage.src,
  },
  {
    id: "3",
    title: "コノスル カルメネール ビシクレタ・レゼルバ",
    chara: "酸っぱい",
    tags: "樽熟成,濃厚",
    image_src: dummyImage.src,
  },
];

export const categoryDummyData: CategoryProps[] = [
  {
    id: 1,
    category: "sober",
    chara: "渋い",
    sub: "red",
  },
  {
    id: 2,
    category: "tart",
    chara: "酸っぱい",
    sub: "red",
  },
  {
    id: 3,
    category: "fruity",
    chara: "果実味豊か",
    sub: "red",
  },
  {
    id: 4,
    category: "white",
    chara: "辛口白",
    sub: "white",
  },
  {
    id: 5,
    category: "sweet",
    chara: "甘口白",
    sub: "white",
  },
  {
    id: 6,
    category: "rose",
    chara: "ロゼ",
    sub: "interim",
  },
  {
    id: 7,
    category: "orange",
    chara: "オレンジ",
    sub: "interim",
  },
  {
    id: 8,
    category: "sparkling",
    chara: "スパークリング",
    sub: "other",
  },
  {
    id: 9,
    category: "fortified",
    chara: "酒精強化",
    sub: "other",
  },
];

export const selectDummyData = categoryDummyData.map((category) => {
  return { value: category.category, label: category.chara };
});

export const wineDummyData: WineProps = {
  id: "sdfaf",
  title: "コノスル カベルネソーヴィニヨン ビシクレタ・レゼルバ",
  author_id: "fsdaf",
  author_name: "ワイン大好き太郎",
  image_src: dummyImage.src,
  price: 2000,
  categories: categoryDummyData[0],
  tags: "樽熟成",
  sober_or_sweet: 3,
  tart: 3,
  fruity: 3,
  description:
    "チリのワインといえばコノスル！\nカベルネソーヴィニヨンの中でも特におすすめの一本です。",
  place: "スーパー,ネット通販",
  erudition:
    "カベルネソーヴィニヨンはボルドー地方の赤ワインの原料としても有名です。",
  category_id: 1,
  created_at: "2021-10-10",
};

export const articleDummyData: ArticleProps = {
  title: "コノスル カベルネソーヴィニヨン ビシクレタ・レゼルバ",
  author_name: "ワイン大好き太郎",
  tags: "樽熟成",
  sober_or_sweet: 3,
  tart: 3,
  fruity: 3,
  categories: categoryDummyData[0],
  id: "1",
  price: 1500,
};
