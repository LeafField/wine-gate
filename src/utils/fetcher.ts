import { supabase } from "./supabase";
import { EditingPageSchemaType, ImageInputSchemaType } from "./schema";
import { User } from "../store";
import { revalidateServer } from "../server/revalidate";

export const postWine = async ({
  value,
  user,
  image,
}: {
  value: EditingPageSchemaType;
  user: User;
  image: ImageInputSchemaType;
}) => {
  if (!image) throw new Error("画像が必須です");
  const IMAGE_SRC = `${user.id}/${image.name}`;

  const { error: imageError, data: imageData } = await supabase.storage
    .from("wines")
    .upload(IMAGE_SRC, image);

  if (imageError) throw new Error("画像のアップロードに失敗しました");
  if (imageData) {
    const { data, error } = await supabase
      .from("wines")
      .insert({
        title: value.title.replace(/　/g, " "),
        price: value.price,
        place: value.place,
        description: value.description,
        erudition: value.erudition,
        category_id: Number(value.category_id),
        fruity: value.fruity,
        tart: value.tart,
        sober_or_sweet: value.sober_or_sweet,
        tags: value.tags.join(),
        author_id: user.id,
        author_name: user.username,
        image_src: imageData.path,
      })
      .select();
    if (error) throw new Error("投稿に失敗しました");
    return data[0].id;
  }
};

export const updateWine = async ({
  id,
  image,
  oldWineImage,
  user_id,
  value,
}: {
  id: string;
  user_id: string;
  value: EditingPageSchemaType;
  image: ImageInputSchemaType | null;
  oldWineImage: string;
}) => {
  revalidateServer(id);
  if (image) {
    const IMAGE_SRC = `${user_id}/${image.name}`;
    const { error } = await supabase.storage
      .from("wines")
      .upload(IMAGE_SRC, image);
    if (error) throw new Error("画像のアップロードに失敗しました");
    const { error: deleteError } = await supabase.storage
      .from("wines")
      .remove([oldWineImage]);
    if (deleteError) throw new Error("画像の削除に失敗しました");
  }
  const { data, error } = await supabase
    .from("wines")
    .update({
      title: value.title.replace(/　/g, " "),
      price: value.price,
      place: value.place,
      description: value.description,
      erudition: value.erudition,
      category_id: Number(value.category_id),
      fruity: value.fruity,
      tart: value.tart,
      sober_or_sweet: value.sober_or_sweet,
      tags: value.tags.join(),
      image_src: image ? `${user_id}/${image.name}` : oldWineImage,
    })
    .eq("id", id)
    .select();
  if (error) throw new Error("ワインの更新に失敗しました");
  return data[0].id;
};

export const favoriteCount = async (wine_id: string) => {
  const { count, error } = await supabase
    .from("favorite")
    .select("id", { count: "exact", head: true })
    .eq("wine_id", wine_id);
  if (error) throw new Error("お気に入りのカウントに失敗しました");
  return count;
};

export const favoriteUserOne = async (wine_id: string) => {
  const { data, error } = await supabase
    .from("favorite")
    .select("user_id")
    .eq("wine_id", wine_id);
  if (error) throw new Error("お気に入りのユーザー取得に失敗しました");
  return data[0] ? data[0] : null;
};

export const addFavorite = async ({
  wine_id,
  user_id,
}: {
  wine_id: string;
  user_id: string;
}) => {
  const { error } = await supabase
    .from("favorite")
    .insert({ wine_id, user_id });

  if (error) throw new Error("お気に入りの登録に失敗しました");
};

export const removeFavorite = async ({
  wine_id,
  user_id,
}: {
  wine_id: string;
  user_id: string;
}) => {
  const { error } = await supabase
    .from("favorite")
    .delete()
    .eq("wine_id", wine_id)
    .eq("user_id", user_id);

  if (error) throw new Error("お気に入りの解除に失敗しました");
};

export const getWine = async (id: string) => {
  const { data, error } = await supabase
    .from("wines")
    .select("*,categories(*)")
    .eq("id", id)
    .single();
  if (error) throw new Error("ワインの取得に失敗しました");

  return data;
};
