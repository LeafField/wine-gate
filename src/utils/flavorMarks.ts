import { ReactNode } from "react";

type Marks = {
  value: number;
  label: ReactNode;
};

const sweet: Marks[] = [
  {
    value: 1,
    label: "辛口",
  },
  {
    value: 2,
    label: "中辛",
  },
  {
    value: 3,
    label: "中口",
  },
  {
    value: 4,
    label: "中甘",
  },
  {
    value: 5,
    label: "甘口",
  },
];

export const tart: Marks[] = [
  {
    value: 1,
    label: "ない",
  },
  {
    value: 2,
    label: "弱い",
  },
  {
    value: 3,
    label: "普通",
  },
  {
    value: 4,
    label: "強め",
  },
  {
    value: 5,
    label: "酸っぱい",
  },
];
