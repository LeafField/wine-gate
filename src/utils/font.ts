import { Noto_Sans_JP, Kumar_One } from "next/font/google";

export const notoStyle = Noto_Sans_JP({ subsets: ["latin"] });
export const KumarOne = Kumar_One({ subsets: ["latin"], weight: "400" });
