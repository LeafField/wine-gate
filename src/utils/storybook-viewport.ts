export const mobileViewport = {
  viewport: {
    defaultViewport: "iphone6",
  },
};
