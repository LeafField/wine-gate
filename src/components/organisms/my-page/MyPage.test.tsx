import { screen, render } from "@testing-library/react";
import MyPage from "./MyPage";
import { wineTitleDummyData } from "../../../utils/dummyData";

jest.mock("next/navigation");

describe("MyPageの単体テスト", () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });
  afterAll(() => {
    jest.restoreAllMocks();
  });

  test("MyPageコンポーネントが存在することを確認する", () => {
    const { container } = render(<MyPage articles={wineTitleDummyData} />);
    expect(screen.getByText(wineTitleDummyData[0].title)).toBeInTheDocument();
    expect(screen.getByText(wineTitleDummyData[1].title)).toBeInTheDocument();
    expect(screen.getByText(wineTitleDummyData[2].title)).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });
});
