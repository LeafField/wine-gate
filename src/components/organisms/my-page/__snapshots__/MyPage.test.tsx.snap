// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MyPageの単体テスト MyPageコンポーネントが存在することを確認する 1`] = `
<div>
  <main
    class="max-w-[64.9375rem] space-y-4 px-4 main tablet:w-[calc(var(--vw)-401px)] tablet:px-0"
  >
    <a
      class="block border border-gray px-2 py-4 transition-colors duration-300 hover:border-blue"
      href="/article/1"
    >
      <div
        class="line-clamp-2 h-full w-full text-heading3"
      >
        コノスル カベルネソーヴィニヨン ビシクレタ・レゼルバ
      </div>
    </a>
    <a
      class="block border border-gray px-2 py-4 transition-colors duration-300 hover:border-blue"
      href="/article/2"
    >
      <div
        class="line-clamp-2 h-full w-full text-heading3"
      >
        コノスル シャルドネ ビシクレタ・レゼルバ
      </div>
    </a>
    <a
      class="block border border-gray px-2 py-4 transition-colors duration-300 hover:border-blue"
      href="/article/3"
    >
      <div
        class="line-clamp-2 h-full w-full text-heading3"
      >
        コノスル カルメネール ビシクレタ・レゼルバ
      </div>
    </a>
  </main>
  <div
    class="footer-area"
  >
    <button
      class="flex items-center justify-center gap-1 transition-colors duration-300 hover:text-blue"
      type="button"
    >
      <svg
        aria-hidden="true"
        class="svg-inline--fa fa-reply mt-[0.125rem]"
        data-icon="reply"
        data-prefix="fas"
        focusable="false"
        role="img"
        viewBox="0 0 512 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M205 34.8c11.5 5.1 19 16.6 19 29.2l0 64 112 0c97.2 0 176 78.8 176 176c0 113.3-81.5 163.9-100.2 174.1c-2.5 1.4-5.3 1.9-8.1 1.9c-10.9 0-19.7-8.9-19.7-19.7c0-7.5 4.3-14.4 9.8-19.5c9.4-8.8 22.2-26.4 22.2-56.7c0-53-43-96-96-96l-96 0 0 64c0 12.6-7.4 24.1-19 29.2s-25 3-34.4-5.4l-160-144C3.9 225.7 0 217.1 0 208s3.9-17.7 10.6-23.8l160-144c9.4-8.5 22.9-10.6 34.4-5.4z"
          fill="currentColor"
        />
      </svg>
      前のページに戻る
    </button>
    <a
      class="flex w-fit items-center justify-center gap-1 transition-colors duration-300 hover:text-blue"
      href="/"
    >
      <svg
        aria-hidden="true"
        class="svg-inline--fa fa-arrow-turn-up mt-[0.125rem]"
        data-icon="arrow-turn-up"
        data-prefix="fas"
        focusable="false"
        role="img"
        viewBox="0 0 384 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M32 448c-17.7 0-32 14.3-32 32s14.3 32 32 32l96 0c53 0 96-43 96-96l0-306.7 73.4 73.4c12.5 12.5 32.8 12.5 45.3 0s12.5-32.8 0-45.3l-128-128c-12.5-12.5-32.8-12.5-45.3 0l-128 128c-12.5 12.5-12.5 32.8 0 45.3s32.8 12.5 45.3 0L160 109.3 160 416c0 17.7-14.3 32-32 32l-96 0z"
          fill="currentColor"
        />
      </svg>
      トップページへ戻る
    </a>
  </div>
</div>
`;
